#################################################################################
#                         STEP 2: ALIGNMENT WITH STAR 
#################################################################################

# Summary
#---------

# The reads from the AC58 project has been trimmed and followed
# by a quality control check. The next step is to map sequence 
# reads to a known transcriptome or annotated genome, transforming 
# the reads into genomic coordinates.
# 
# The reads will be mapped using STAR which is a genome reference 
# based aligner. To run STAR two steps must be followed:
#  - Step 1. Build STAR genome index (00_STAR_genome_index.sh)
#  - Step 2. Mapping reads with STAR is performed in this script
#
# The reference genome for the read length 101 based on the genome 
# assembly hg38 of 2020 which is compatible with GRCh38, and the 
# corresponding genome annotation for the 2-pass mapping. The 2-pass 
# mapping aims to map novel splice junctions.
# 
# The output files are mapped and unmapped reads in SAM format,
# genes counts, transcripts in SAM format and the splice junctions 
# sites, as well, as the log files.

# Requirement
# ---------------------
#
# We use the genome index generated by Rubén in folder Human_101.
# If it was not prepared, we should have run the script 03_STAR_genome_index


###############################################################
#               Experimental Information
###############################################################

# Project: AC58

# The raw data analyzed is contained in /vols/GPArkaitz_bigdata/DATA_shared/AC58
# Human cell line
# Total bulk RNAseq paired-end
# The inserts size 40 - 300 bp
# Library average size 327 bp
# Read length 101
# Library kit TruSeq Stranded Total RNA



###############################################################
#                       STAR Version
###############################################################

# R version is the latest in March 2023
# R version 4.3.0 (2023-04-21 ucrt)

# STAR version 2.7.10b


###############################################################
#                       STAR Manuals
###############################################################

## Manuals used 

# STAR manual 2.7.10b
# October 20, 2022
# Link: https://github.com/alexdobin/STAR


# STAR manual 2.7.0a
# January 23, 2019
# Link: https://physiology.med.cornell.edu/faculty/skrabanek/lab/angsd/lecture_notes/STARmanual.pdf



###############################################################
#                   Parameters selection
###############################################################


# Run parameters
#     --runThreadN
#         Number of threads used for the alignment, we use the 
#         number of cores available. 
# 
# Genome parameters
#     --genomeDir
#         Path to the genome indexes directory
#         PATH = /vols/GPArkaitz_bigdata/mponce/AC35/02_STAR/Mouse_101
# 
# Read parameters
#     --readFilesIn 
#         Files with the path. The files contain the reads to 
#         align. We used paired-end reads so two files are given 
#         with their specific path.
#         PATH = /vols/GPArkaitz_bigdata/mponce/AC35/01_TRIMMED/PolyG
# 
#     --readFilesCommand Uncompressed the fastq files. In this 
#         case, we use the command zcat.
# 
# Limits parameters
#     --limitBAMsortRAM 
#         Maximum available RAM (bytes) for sorting BAM. If =0, it 
#         will be set to the genome index size. The value is: 
#         10 x Genome size (GB).
# 
#         We use 30 GB of RAM. If not the following error will appear:
# 
#         EXITING: fatal error trying to allocate genome arrays, exception thrown: std::bad_alloc
#         Possible cause 1: not enough RAM. Check if you have enough RAM 28703634434 bytes
#         Possible cause 2: not enough virtual memory allowed with ulimit. SOLUTION: run ulimit -v 28703634434
# 
# Output parameters
#     --outSAMtype 
#         Specify the format of the output files if SAM/BAM, sorted 
#         by coordinates, ... 
#         Default option is SAM
#         We choose SAM
# 
#     --outReadsUnmapped 
#         Specifies that unmapped reads will be given in a separate 
#         file from th euniqueliy mapped reads. 
#         We use Fastx option.
# 
#     --outFileNamePrefix
#         Define the output path and name of the folder
#         We use STAR_samplename
# 
#     --outFilterMultimapNmax 
#         default: 10
#         Maximum number of loci the read is allowed to map to. 
#         Alignments (all of them) will be output only if the read maps 
#         to no more loci than this value. Otherwise no alignments will be output, 
#         and the read will be counted as "mapped to too many loci" in the Log.final.out 
# 
#         If --outFilterMultimapNmax  1, limits the output to just the uniquely-mapping 
#         reads (i.e. reads that confidently map to only one locus). If the genes of 
#         interest have high sequence similarity, this option will probably eliminate 
#         a large number of reads that map to multiple loci.
# 
#         We chose --outFilterMultimapNmax  1
# 
# Quantification of Annotations
#     --quantMode 
#         The type of quantification request. In our case, we want 
#         the gene counts and the transcripts, so we chose:
#         TranscriptomicSAM GeneCounts
# 
#         The result of the GeneCounts in Mapped only includes the 
#         Uniquely mapped reads.
# 
# 2/pass Mappings
#     --twopassMode 
#         We used Basic which perform a basic 2-pass mapping, with 
#         all 1st pass junctions inserted into the genome indices on 
#         the fly. 
#
#
####################### CONSIDERATIONS #######################
#
#
# Consideration 1
#-----------------
# 
# The parameter outFilterMultimapNmax only applies to genomic alignments 
# which means that when mapping the reads towards the genome of reference
# it limits the number of read that can be multimapped. In our case, we chose
# 1, so only uniquely mapped read are chosen.
# 
# This parameter does not affect GeneCounts, but it was used in previous scripts
# so we will maintain this parameter because it can be of interest in the future.
# 
# Situation of interest (i.e. paralog genes, pseudogenes, ...)
# 
# If a read maps perfectly to one location and to another with a couple of 
# mismatches, such as in the situation where the read is from a gene, but 
# has similarity to a pseudogene, is it possible to control alignments reported 
# in terms of their optimality ? From my reading of the user manual, STAR would 
# report both alignments as long as they meet the user's multiple mapping location 
# count thresholds.
#
# These alignments are not output to Aligned.out.sam if Nmap > outFilterMultimapNmax.
#
# If you wan to get just the multimappers, the best way is to allow them (i.e. 
# use --outFilterMultimapNmax 100), and then filter them from the sam file by 
# the MAPQ field: the unique mappers will have MAPQ=255, while multimappers will 
# have MAPQ<255. Also, unique mappers have NH:i:1 attribute, while multimappers 
# have NH:i:<Nmult>, where Nmult is the number of loci they map to.
# 
#
# Consideration 2
#-----------------
# 
# The SPLICING JUNCTIONS output files includes all reads which means 
# that includes the uniquely mapped and the multimapping reads.
# If you want to modify the output used the parameter --outSJfilterReads 
# with the option Unique. 
#
#
# Consideration 3
#-----------------
# 
# Reason why we performed SAM instead of sorted by coordinates BAM format. 
# The --limitBAMsortRAM parameter showed the following error:
# 
#         EXITING because of fatal ERROR: not enough memory for BAM sorting: 
#         SOLUTION: re-run STAR with at least --limitBAMsortRAM 1288754585
#         May 09 16:46:23 ...... FATAL ERROR, exiting
#         
# The solution found was to remove this parameter, which will turn into the 
# default value, 0, which would take the genome indexes size as reference. 
# This solution make sense to me because the alignment was not affected and 
# BAM file will not be used. 
# 
# Finally, we decided to maintain the parameter and change the output to SAM
# format.





###############################################################
#                       Output
###############################################################

## Log files 

# Log.out Detail information about the run 

# Log.progress.out Progress job information updated each minute. 
#     Consult while running 

# Log.final.out Summary mapping statistics after mapping job is 
#     complete. The parameters of interest are: 
#       - Uniquely mapped reads % which should be always over 80%.
#         90% very good library.
#       - % of reads mapped to multiple loci which give information 
#         about the multimapping. 
#       - % of reads unmapped: which gives information about the 
#         cause of the mismatches.


## Mapped reads with BAM format (Aligned.sortedByCoord.out.bam)
# Binary BAM format, thus saving time on converting SAM files to BAM. 
# It can also sort BAM files by coordinates, which is required by 
# many downstream applications.


## Unmapped reads (Unmapped.out.mate1 and Unmapped.out.mate2)

# For paired-end reads, if a read maps as a whole, but one of the mates 
# does not map, both mates will also be output in their corresponding file. 
# To indicate the mapping status of the read mates,the following tags are 
# appended to the read name:
#   00: mates were not mapped;
#   10: 1st mate mapped, 2nd unmapped
#   01: 1st unmapped, 2nd mapped


## Gene Counts (ReadsPerGene.out.tab)

# Column 1. Gene ID
# Column 2. counts for unstraned RNA-seq
# Column 3. counts for 1st read strand aligned with RNA
# Column 4. counts for 2nd read strand aligned with RNA 


## Splice junctions (SJ.out.tab)
# High confidence collapsed splice junctions in tab-delimited format. 
# Note that STAR defines the junction start/end as intronic bases, 
# while many other software define them as exonic bases.

# Column 1. Chromosome
# Column 2. first base of the intron
# Column 3. last base of the intron
# Column 4. strand. 1:+ / 2:-
# Column 5. intron
# Column 6. Note that in 2-pass mode, junctions detected in the 1st pass 
#           are reported as annotated (1), in addition to annotated 
#           junctions from GTF. Unannotated: 0
# Column 7. Number of uniquely mapping reads crossing the junction
# Column 8. Number of multi-mapping reads crossing the junction
# Column 9. Maximum spliced alignment overhang. It should be 10 


## Transcript coordinates (Aligned.toTranscriptome.out.bam)

# These transcriptomic alignments can be used with various transcript 
# quantification software that require reads to be mapped to transcriptome, 
# such as RSEM or eXpress. 

# Note, that STAR first aligns reads to entire genome, and only then searches 
# for concordance between alignments and transcripts.This approach offers 
# certain advantages compared to the alignment to transcriptome only, by not 
# forcing the alignments to annotated transcripts. Note that --outFilterMultimapNmax
# only applies to genomic alignments. If an alignment passes this filter, it 
# is converted to all possible transcriptomic alignments and all of them are output.



###############################################################
# How to run in INDAR
# 
# 1. Input folder is inside my user folder in BigData
#
# 2. Output folder is inside my user folder in BigData
#
# 3. This script will generate all the "PBS" file, one per each 
# sample, which includes the commands to successfully run STAR. 
# The PBS are directly send to the job scheduler (SLURM) to run 
# the jobs in INDAR.
#
###############################################################


### Access R ###
source /opt/ohpc/pub/apps/R/R-4.2.1/cic-R
R
# Load R libraries 
.libPaths("/vols/GPArkaitz_bigdata/DATA_shared/NewCluster_R")



### General Project ###
# Project Name 
project_name <- "AC58"

# Output and input directories SSH 
dir_infiles <- paste("/vols/GPArkaitz_bigdata/mponce/", project_name, "/05_DEGs/01_TRIMMED", sep = "" )
dir_outfiles <- paste("/vols/GPArkaitz_bigdata/mponce/", project_name, "/05_DEGs", sep = "")

# Create output directory
dir.create(file.path(dir_outfiles,"03_STAR"))
dir_outfiles <- paste(dir_outfiles,"/03_STAR",sep='')
setwd(dir_outfiles)

# Set directory
setwd(dir_outfiles)


### Trimmed FASTQs ###
# List trimmed fastq.gz files
samples <- list.files(path=dir_infiles, pattern = "_1_trmd.fastq.gz")
# Filter sample name 
samples_names <- gsub("_1_trmd.fastq.gz","",samples)

# Genome indexes for reads with length 101
genome_dir <- paste("/vols/GPArkaitz_bigdata/mponce/", project_name, "/05_DEGs/03_STAR/Human_101", sep = "")


### Process information ###
partition <- "FAST"
time <- c("03:00:00")
memory <- c("40")
node <- 1
cpu <- 6
ram <- as.numeric(memory)*10^9


###############################################################
#                           STAR
###############################################################

### Generate PBS ###
for (i in 1:length(samples_names)) {
  # Job name
  job_name <- paste(samples_names[i],"_STAR",sep='');
  
  # Input files
  input1 <- paste(samples_names[i],"_1_trmd.fastq.gz", sep="")
  input2 <- paste(samples_names[i],"_2_trmd.fastq.gz", sep="")
  
  # Output files in this directory
  output <- paste(dir_outfiles, "/", job_name, "/", sep="")
  
  # Program Path 
  path <- c("source  /opt/ohpc/pub/apps/star/STAR-2.7.10a/cic-env")
  
  # Command
  command <- paste("STAR --genomeDir", genome_dir,             
                   "--runThreadN", cpu,                         
                   "--readFilesCommand zcat",                  
                   "--readFilesIn", input1, input2,   
                   "--limitBAMsortRAM", ram,
                   "--twopassMode Basic ",                     
                   "--quantMode TranscriptomeSAM GeneCounts",
                   "--outSAMtype SAM",
                   "--outReadsUnmapped Fastx",                 
                   "--outFileNamePrefix", output,              
                   "--outFilterMultimapNmax 1",                
                   sep=" ") 
   
  # SBATCH File
  filename <- paste(job_name,".sh",sep='');
  cat(
    c("#!/bin/sh"),
    paste("#SBATCH --job-name=",job_name,sep=''),
    paste("#SBATCH --partition=",partition,sep=''),
    c("#SBATCH --ntasks=1"),
    paste("#SBATCH --nodes=", node,sep =''), 
    paste("#SBATCH --cpus-per-task=",cpu,sep=''),
    paste("#SBATCH --time=",time,sep=''),
    paste("#SBATCH --mem=",memory,"GB",sep=''),
    paste("#SBATCH -o ",job_name,".out",sep=''),
    paste("#SBATCH -e ",job_name,".err",sep=''),
    # c("#SBATCH --exclude=gn[10-13]"),
    c("#SBATCH --mail-type=FAIL,END"),
    c("#SBATCH --mail-user=mponce@cicbiogune.es"),
    "\n\n",
    c(path), 
    "\n",
    c(paste("cd ", dir_infiles)), 
    "\n",
    c(command),
    file=filename,sep = "\n",append=F)
  
  # Run PBS
  system(paste("sbatch",filename,sep=' '));
  Sys.sleep(3)
}



q()

